### PR TITLE
New target KFC32F3_INAV

### PR DESCRIPTION
**Pull request by marbalon**
Just another target, F3 board with:
- MPU6000
- MS5611
- OSD
- build in RC - eLeReS - quite popular in Poland
- SPI flash
- USB - VCP
- 2x UART, 1x SPI, 10xPWM, RSSI, ADC
http://4drones.pl/files/KFC32F3_inav_top.jpg

### CODE REVIEW

**Comment by marbalon** [comment]
If this is possible merge it before 1.7.1, changes not related with other targets.
Test failed because of NAZE flash limit...

**Comment by Linjieqiang** [comment]
What SPI receiver?

File: src/main/config/parameter_group_ids.h [hunk 1]
```diff
@@ -73,7 +73,8 @@
 #define PG_SERVO_CONFIG 52
 //#define PG_IBUS_TELEMETRY_CONFIG 53
 //#define PG_VTX_CONFIG 54
-#define PG_CF_END 52
+#define PG_ELERES_CONFIG 54
```
Comment (DzikuVx) [inline src/main/config/parameter_group_ids.h]: @marbalon clash ?

File: src/main/sensors/barometer.c [hunk 2]
```diff
@@ -260,9 +259,9 @@ uint32_t baroUpdate(void)
         case BAROMETER_NEEDS_CALCULATION:
             baro.dev.get_up();
             baro.dev.start_ut();
-            baro.dev.calculate(&baroPressure, &baro.baroTemperature);
+            baro.dev.calculate(&baro.baroPressure, &baro.baroTemperature);
```
Comment (DzikuVx) [inline src/main/sensors/barometer.c]: This somehow looks like is affecting current baro implementation. I suppose it should be tested on other boards

**Comment by kabturek** [comment]
Great stuff, looking forward to using the board with an official release!

**Comment by digitalentity** [comment]
What's eLeReS receiver? Can we use the same code for REVO's RFM module to
have built-in receiver as well?

**Comment by marbalon** [comment]
Probably yes. if it is connected via SPI to main CPU. But this is compatible only with eLeReS RC - it is closed source (this receiver part is open) link quite popular in Poland.

**Comment by digitalentity** [comment]
If receiver part is open, we can implement a tx part that talks the same
protocol no problem.

File: src/main/config/parameter_group_ids.h (hunk 1 above)
Comment (DzikuVx) [reply src/main/config/parameter_group_ids.h]: my bad

**Review by DzikuVx — APPROVED** [review]

File: src/main/config/parameter_group_ids.h (hunk 1 above)
Comment (digitalentity) [inline src/main/config/parameter_group_ids.h]: Value of 54 is reserved for `PG_VTX_CONFIG`

File: src/main/sensors/barometer.c (hunk 2 above)
Comment (digitalentity) [inline src/main/sensors/barometer.c]: This changes behavior of baro code. Was is tested to not break things on other boards?

File: src/main/fc/cli.c [hunk 3]
```diff
@@ -152,7 +153,7 @@ static const char * const featureNames[] = {
     "", "TELEMETRY", "CURRENT_METER", "3D", "RX_PARALLEL_PWM",
     "RX_MSP", "RSSI_ADC", "LED_STRIP", "DASHBOARD", "",
     "BLACKBOX", "CHANNEL_FORWARDING", "TRANSPONDER", "AIRMODE",
-    "SUPEREXPO", "VTX", "RX_SPI", "SOFTSPI", "PWM_SERVO_DRIVER", "PWM_OUTPUT_ENABLE", "OSD", NULL
+    "SUPEREXPO", "VTX", "RX_SPI", "SOFTSPI", "PWM_SERVO_DRIVER", "PWM_OUTPUT_ENABLE", "OSD", "RX_ELERES", NULL
```
Comment (digitalentity) [inline src/main/fc/cli.c]: Looks like RX_ELERES is similar to RX_SPI. I think it should be implemented using RX_SPI infrastructure - we don't need two separate implementations of SPI-based receivers.

**Review by digitalentity — CHANGES_REQUESTED** [review]

**Comment by digitalentity** [comment]
I will gladly merge KFC32F3 without eLeReS support into 1.7.1. I think we need to have eLeReS receiver support under RX_SPI infrastructure (similar to nRF24 receivers). Also, low-level hardware-related code (driving RFM module) should be moved to `drivers` directory - `rx` directory should be hardware-independent.

**Comment by martinbudden** [comment]
I agree, eLeReS should be updated to use the RX_SPI framework.

File: src/main/sensors/barometer.c (hunk 2 above)
Comment (marbalon) [reply src/main/sensors/barometer.c]: Yes, it was tested. just move global variable in to baro struct.

File: src/main/fc/cli.c (hunk 3 above)
Comment (marbalon) [reply src/main/fc/cli.c]: Yes, it is stil similar to other SPI RX but require IRQ check in separate task etc. and also other configuration like telemetry, telemetry power etc. But will try to code it as regular RX_SPI receiver.

Comment (martinbudden) [reply src/main/fc/cli.c]: You don't need IRQ check in a separate task - the framework already provides for that. `rxUpdateCheck` in `rx.c` is called regularly. This in turn calls `rxRuntimeConfig.rcFrameStatusFn()` a function provided by the receiver implementation. For SPI this is `rxSpiFrameStatus` in `rx_spi.c`, which in turn calls the function pointer `protocolDataReceived` which is the thing that you should provide and set up. See `rxSpiSetProtocol` for how other protocols set this up.

Comment (marbalon) [reply src/main/fc/cli.c]: Yes, but it is called in TASK_RX, at 50Hz rate... but I found that IRQ functions for this link need to be called more often, like 500Hz.

Comment (digitalentity) [reply src/main/fc/cli.c]: `rxUpdateCheck` is not a task - it's a task check function which is called each iteration of scheduler (several kHz rate).

**Comment by digitalentity** [comment]
We need to do something about NAZE target :thinking:

**Comment by digitalentity** [comment]
@marbalon please remove the eleres source from this PR (you can add it later at any time).

File: src/main/fc/cli.c (hunk 3 above)
Comment (marbalon) [reply src/main/fc/cli.c]: Ok, will try it.

Comment (martinbudden) [reply src/main/fc/cli.c]: To confirm what @digitalentity says, `rxUpdateCheck` is not called from within `TASK_RX`, it's called from the main scheduler loop. This means your `protocolDataReceived` should just check if the interrupt has occurred and return with the appropriate value. The receiver data should be processed in the `protocolSetRcDataFromPayload` function.

**Review by digitalentity — APPROVED** [review]

File: src/main/target/KFC32F3_INAV/hardware_revision.c [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+#include "drivers/sensor.h"
+#include "drivers/io.h"
+#include "drivers/gpio.h"
+#include "drivers/exti.h"
+
+#include "hardware_revision.h"
+
+uint8_t hardwareRevision = 1;
+
+void detectHardwareRevision(void)
```
Comment (martinbudden) [inline src/main/target/KFC32F3_INAV/hardware_revision.c]: This function is a misnomer, in that it seems to have nothing to do with detecting the hardware revision. It's also a non-standard way of setting up the GPIO pins.

**Review by martinbudden — CHANGES_REQUESTED** [review]
Please standardise `detectHardwareRevision `

File: src/main/target/KFC32F3_INAV/hardware_revision.c (hunk 4 above)
Comment (marbalon) [reply src/main/target/KFC32F3_INAV/hardware_revision.c]: There is no early_setup() function in inav/BF code, but sometimes if you have board with a lot of SPI devices you need to force all CS signals to high, to be sure that sensors will be detected correctly. Without this sometimes board have a problem with initialization.

Comment (digitalentity) [reply src/main/target/KFC32F3_INAV/hardware_revision.c]: Actually there is. 
Function is called `void initialisePreBootHardware(void)` and it's controlled by `USE_HARDWARE_PREBOOT_SETUP` define. See `PIXRACER` target.

**Review by martinbudden — APPROVED** [review]

**Comment by digitalentity** [comment]
@marbalon merged.

**Comment by digitalentity** [comment]
Looking forward to have eLeReS support. Looks like we can also have OpenLSR support on REVO and Sparky2

**Comment by martinbudden** [comment]
Yes, I also look forward to seeing eLeReS support. I think it should be fairly straightforward to fit it into the RX framework.

**Comment by marbalon** [comment]
I checked RX stuff a little and hardware part can be removed and rx_spi.c from driver can be used. But I'm afraid eLeReS can't be another RX_SPI because eLeReS + RFM22 works different than NRF24 based modules.

btw.

**Comment by martinbudden** [comment]
I think the best approach is for you to use the framework as much as possible, and if changes are required in the framework we can add them. The RX_SPI framework was initially developed for NRF24, but is not meant to be exclusively for NRF24. For example there is ongoing work to support  CC2500. Generally when a new class of items is added to a framework, that framework may need changes, so this is all par for the course.

What are the problems the existing framework causes for RFM22?

**Comment by marbalon** [comment]
> What are the problems the existing framework causes for RFM22?
I will made a separate branch for this and will try to change eLeReS driver as much as I can to existing API and then we will talk letter... Now we can close this branch.

**Comment by Linjieqiang** [comment]
@marbalon I saw your target code.You shared the SPI2 bus with MAX7456,MPU6000,RX_RFM22,FLASH.Was  it working well?

**Pull request merged**